app/jobs/github_commits_job.rb

```diff
@@ -55,12 +55,9 @@ def update_user_commit_counts_for(project)
     end
 
   def get_contributor_stats(repo)
-    loop do
-      contributors = client.contributor_stats(repo)
-      return contributors unless contributors.nil?
-      Rails.logger.warn "Waiting for Github to calculate project statistics for #{repo}"
-      sleep 3
-    end
+    contributors = client.contributor_stats(repo, { retry_timeout: 6, retry_wait: 3 })
```

Comment:
@praveendhawan @tansaku as it can be seen [here](https://github.com/octokit/octokit.rb/edit/master/lib/octokit/client/stats.rb):

```
module Octokit
  class Client

    # Methods for the Repository Statistics API
    #
    # @see https://developer.github.com/v3/repos/statistics/
    module Stats

      # Get contributors list with additions, deletions, and commit counts
      #
      # @param repo [Integer, String, Hash, Repository] A GitHub repository
      # @option retry_timeout [Number] How long Octokit should keep trying to get stats (in seconds)
      # @option retry_wait [Number] How long Octokit should wait between retries.
      # @return [Array<Sawyer::Resource>] Array of contributor stats
      # @see https://developer.github.com/v3/repos/statistics/#get-contributors-list-with-additions-deletions-and-commit-counts
      # @example Get contributor stats for octokit
      #   @client.contributors_stats('octokit/octokit.rb')
      def contributors_stats(repo, options = {})
        get_stats(repo, "contributors", options)
      end
      alias :contributor_stats :contributors_stats
```

On the implementation of the method `contributor_stats` we can send those two parameters for making the retries and the time that it should wait making retries. I just did the setup for the request to `timeout` in 6 seconds, this means that the it will retry for a maximum period of 6 seconds. I also did the setup for how longer octokit should wait between retries to 3 seconds. This setup was totally arbitrary and based on what I considered it could be a good configuration, but I'm maybe wrong and I would really appreciate your opinion here :+1:

Reply:
here you have the definition for the method `get_stats`:

```ruby

      private

      # @private Get stats for a repository
      #
      # @param repo [Integer, String, Hash, Repository] A GitHub repository
      # @param metric [String] The metrics you are looking for
      # @return [Array<Sawyer::Resource> or nil] Stats in metric-specific format, or nil if not yet calculated.
      # @see https://developer.github.com/v3/repos/statistics/
      def get_stats(repo, metric, options = {})
        if retry_timeout = options.delete(:retry_timeout)
          retry_wait = options.delete(:retry_wait) || 0.5
          timeout = Time.now + retry_timeout
        end
        loop do
          data = get("#{Repository.path repo}/stats/#{metric}", options)
          return data if last_response.status == 200
          return nil unless retry_timeout
          return nil if Time.now >= timeout
          sleep retry_wait if retry_wait
        end
      end
```

Reply:
nice findings @nisevi . Looks good to me as well @tansaku. But I would say we should have that test case for empty repo as well.